core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java

```diff
@@ -98,22 +112,57 @@ public void processExpiration() {
                      expiredMortal = ExpiryHelper.isExpiredMortal(lifespan, ice.getCreated(), currentTimeMillis);
                      expiredTransient = ExpiryHelper.isExpiredTransient(maxIdle, ice.getLastUsed(), currentTimeMillis);
                   }
-                  // We check lifespan first as this is much less expensive to remove than max idle.
-                  if (expiredMortal) {
-                     addAndWaitIfFull(handleLifespanExpireEntry(ice.getKey(), value, lifespan, false), futures);
-                  } else if (expiredTransient) {
-                     addAndWaitIfFull(actualRemoveMaxIdleExpireEntry(ice.getKey(), value, maxIdle, false), futures);
+                  if (expiredMortal || expiredTransient) {
+                     if (expirationPermits.tryAcquire(1, TimeUnit.MINUTES)) {
+                        removedEntries++;
+                        CompletableFuture<?> stage;
+                        // We check lifespan first as this is much less expensive to remove than max idle.
+                        if (expiredMortal) {
+                           stage = handleLifespanExpireEntry(ice.getKey(), value, lifespan, false);
+                        } else {
+                           stage = actualRemoveMaxIdleExpireEntry(ice.getKey(), value, maxIdle, false);
+                        }
+                        futures.add(stage);
+                        stage.whenComplete((t, obj) -> {
+                           if (trace) {
+                              log.tracef("There was a problem expiring the entry for key: " + ice.getKey(), t);
+                           }
+                           expirationPermits.release();
+                           futures.remove(stage);
+                        });
+                     } else {
+                        if (trace) {
+                           log.tracef("Purging data container stopped due to waiting for prior removal (could be a bug or misconfiguration). Total time was: %s and removed %d entries",
+                                 Util.prettyPrintTime(timeService.timeDuration(start, TimeUnit.MILLISECONDS)), removedEntries);
+                        }
+                        return;
+                     }
                   }
                }
-            });
+               // Short circuit if topology has changed
+               if (distributionManager.getCacheTopology() != topology) {
+                  if (trace) {
+                     log.tracef("Purging data container stopped due to topology change. Total time was: %s and removed %d entries",
```

Comment:
Maybe we should start a new purge cycle immediately after a topology change, not just stop the current cycle?

Reply:
Hrmm, I guess we could. It would be more helpful when the reaper wakeup is too long.